### PR TITLE
Add support for precise tainting of byte array objects

### PR DESCRIPTION
**Pull request by manuel-alvarez-alvarez**
# What Does This Do
Adds support for accurate tainting of byte array instances.

# Motivation
We need to keep track of multiple ranges in order to add proper support for Kafka 3.x as they use byte buffers extensively.

### CODE REVIEW

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/propagation/PropagationModuleImpl.java [hunk 1]
```diff
@@ -118,6 +140,45 @@ public void taintIfTainted(
...
+  }
+
+  @Override
+  public void taintIfTainted(
+      @Nullable final IastContext ctx,
+      @Nullable final Object target,
+      @Nullable final Object input,
+      final int start,
+      final int length,
+      boolean keepRanges,
```
Comment (smola) [inline dd-java-agent/agent-iast/src/main/java/com/datadog/iast/propagation/PropagationModuleImpl.java:161]: I'm a bit concerned about the excessive overloading `taintIfTainted` functions, especially adding boolean parameters (not obvious from the call).

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/taint/Ranges.java [hunk 2]
```diff
@@ -217,6 +243,36 @@ public static Range[] newArray(final long size) {
     return new Range[size > MAX_RANGE_COUNT ? MAX_RANGE_COUNT : (int) size];
   }
 
+  /** Insert the new range maintaining order (it assumes that both arrays are already sorted) */
+  public static Range[] insert(final Range[] currentRanges, final Range... newRanges) {
```
Comment (smola) [inline dd-java-agent/agent-iast/src/main/java/com/datadog/iast/taint/Ranges.java:247]:
```suggestion
  public static Range[] append(final Range[] currentRanges, final Range... newRanges) {
```

Can we call this `append` or `concat`? `insert` sounds like an in-place operation.

**Review by smola — APPROVED** [review]
LGTM. Just a couple of stylistic comments.

**Review by jandro996 — APPROVED** [review]

**Review by ValentinZakharov — APPROVED** [review]

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/taint/Ranges.java (hunk 2 above)
Comment (manuel-alvarez-alvarez) [reply dd-java-agent/agent-iast/src/main/java/com/datadog/iast/taint/Ranges.java]: Changed into `mergeRangesSorted` as it makes more sense as in a merge sort algo (also rename the vars to make it clear what it does)

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/propagation/PropagationModuleImpl.java (hunk 1 above)
Comment (manuel-alvarez-alvarez) [reply dd-java-agent/agent-iast/src/main/java/com/datadog/iast/propagation/PropagationModuleImpl.java:161]: Agreed the API is too verbose and makes it difficult to reason about it and of course use it. Can we try to do a refactor in a separate PR?

**Pull request merged**